examples/03_connectivity/plot_rest_clustering.py

```diff
@@ -30,130 +52,144 @@
-# Perform Ward clustering
-# -----------------------
+#########################################################################
+# Brain parcellations with Ward Clustering
+# ----------------------------------------
 #
-# We use spatially-constrained Ward clustering. For this, we need to
-# compute from the mask a matrix giving the voxel-to-voxel connectivity
+# Transforming list of images to data matrix and build brain parcellations,
+# all can be done at once using class `Parcellations`.
 
-# Compute connectivity matrix: which voxel is connected to which
-from sklearn.feature_extraction import image
-shape = mask.shape
-connectivity = image.grid_to_graph(n_x=shape[0], n_y=shape[1],
-                                   n_z=shape[2], mask=mask)
-
-
-##################################################################
-# Then we use FeatureAgglomeration from scikit-learn. Indeed, the voxels
-# are the features of the data matrix.
-#
-# In addition, we use caching. As a result, the clustering doesn't have
-# to be recomputed later.
+from nilearn.regions import Parcellations
 
-# Computing the ward for the first time, this is long...
-from sklearn.cluster import FeatureAgglomeration
-# If you have scikit-learn older than 0.14, you need to import
-# WardAgglomeration instead of FeatureAgglomeration
+# Computing ward for the first time, will be long... This can be seen by
+# measuring using time
 import time
 start = time.time()
-ward = FeatureAgglomeration(n_clusters=1000, connectivity=connectivity,
-                            linkage='ward', memory='nilearn_cache')
-ward.fit(fmri_masked)
+
+# Agglomerative Clustering: ward
+
+# We build parameters of our own for this object. Parameters related to
+# masking, caching and defining number of clusters and specific parcellations
+# method.
+ward = Parcellations(method='ward', n_parcels=1000,
+                     standardize=False, smoothing_fwhm=2.,
+                     memory='nilearn_cache', memory_level=1,
+                     verbose=1)
+# Call fit on functional dataset: single subject (less samples).
+ward.fit(dataset.func)
 print("Ward agglomeration 1000 clusters: %.2fs" % (time.time() - start))
 
-# Compute the ward with more clusters, should be faster as we are using
-# the caching mechanism
+# We compute now ward clustering with more number of clusters=2000 and compare
+# time with 1000 clusters. To see the benefits of caching for second time.
+
+# We initialize class again with n_parcels=2000 this time.
 start = time.time()
-ward = FeatureAgglomeration(n_clusters=2000, connectivity=connectivity,
-                            linkage='ward', memory='nilearn_cache')
-ward.fit(fmri_masked)
+ward = Parcellations(method='ward', n_parcels=2000,
+                     standardize=False, smoothing_fwhm=2.,
+                     memory='nilearn_cache', memory_level=1,
+                     verbose=1)
+ward.fit(dataset.func)
 print("Ward agglomeration 2000 clusters: %.2fs" % (time.time() - start))
```

Comment:
If I run this example, the caching doesn't seem to work: the second call to fit takes the same time as the first call. Maybe it's a question of adjusting how "memory_level" is propagated in the object?

Reply:
For me, I renamed the memory to new cache directory. It looks like it is working.
You have less time taken by object for second time.

Ward agglomeration 1000 clusters: 15.44s
Ward agglomeration 2000 clusters: 10.84s